### ISSUE TITLE
[reportArgumentType] Passing a function with fewer arguments incorrectly errors with "Function accepts too many positional parameters"

### ISSUE THREAD

**Issue by mark-henry01** [microsoft/pyright#8105]
**Describe the bug**
**Code or Screenshots**

Minimal example:
```python
from collections.abc import Callable

def foo(fn: Callable[[str], str]):
    pass

def fn1(s: str) -> str:
    return f"hello {s}"

def fn2() -> str:
    return "hello"

def fn3(i: int) -> str:
    return f"hello number {i}"

def fn4(s: string, i: number) -> str:
    return f"hello {s} and number {i}"

foo(fn1) # works (OK)
foo(fn2) # errors (not OK)
foo(fn3) # errors (OK)
foo(fn4) # errors (OK)
```

A function `f` is a subtype of another function `g`, if arguments of `f` have same or wider type than `g` (neglecting return type for this issue). `f` may also define less parameters than `g`. In more technical terms a function is contravariant in its parameters.

Following this rule, we should be able to pass `fn2` with signature `() -> str` as argument of `foo`, because it defines less parameters than `Callable[[str], str]` / `(str) -> str`. But this is not the case with `foo(fn2)`, which reports:
> Argument of type "() -> str" cannot be assigned to parameter "fn" of type "(str) -> str" in function "foo"
     Type "() -> str" is incompatible with type "(str) -> str"
       Function accepts too many positional parameters; expected 0 but received 1 [reportArgumentType]

If I repeat [same example in TypeScript](https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABMOcAUwwC5FoM455QBOMYA5gJSIC8AfIkaRdQN4C+AUJxAkcmACMtXAUYkyVWgwBEACwCmAGyVxEAdzjElAE0QzEAakY8+UAQCYRaavX2KVcGabD9MAZmswcYEAFsAIwViW1kHVX0jRBgXNzAAFmsxJkkAGmiffyCQ6XtlCINjGM4UdExBShLUDDALStKa93rqzHjKIA) with `strictFunctionTypes` enabled, behavior is as expected, no error is reported for `foo(fn2)`.

From my point of view this is incorrectly flagged by `reportArgumentType` - or do I have false assumptions regarding this rule?

**Comment by LeeeeT** [comment]
In python, if you call a function with more arguments than it needs, it will result in a runtime error. Pyright prevents you from making such a mistake by disallowing assigning `Callable[[], str]` to `Callable[[str], str]`.

Javascript behaves differently when passing too many arguments to a function and typescript reflects this. You can read more [here](https://www.typescriptlang.org/docs/handbook/2/functions.html#optional-parameters-in-callbacks).

**Comment by mark-henry01** [comment]
@LeeeeT Yes you're right, thanks for the quick response! Just had the same thought.

I need to state, this was a special case with FastAPI dependencies, where I needed to pass a factory function with less arguments in context of a test environment (no need for things like remote API etc. here). This function still needed to conform to the function type interface with more args declared. As from test environment I am ensured to not call this function with too many arguments, it's just about suppressing the error notification in IDE. I think it makes sense to ignore the error for this particular case via `# pyright: ignore [reportArgumentType]`.

Hence closing the issue, feel free to reopen otherwise.
